src/lock.rs

```diff
@@ -0,0 +1,328 @@
+    SHA256,
+}
+
+impl std::fmt::Display for DigestAlgorithm {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let s = match self {
+            Self::SHA256 => SHA256_TAG,
+        };
+        f.write_str(s)
+    }
+}
+
+impl FromStr for DigestAlgorithm {
+    type Err = eyre::Report;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            SHA256_TAG => Ok(Self::SHA256),
+            other => eyre::bail!("invalid digest algorithm: {other}"),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct Digest {
+    bytes: Vec<u8>,
+    algorithm: DigestAlgorithm,
+}
+
+impl TryFrom<digest::Digest> for Digest {
+    type Error = eyre::Report;
+
+    fn try_from(value: digest::Digest) -> Result<Self, Self::Error> {
+        let algorithm = if value.algorithm() == &digest::SHA256 {
+            DigestAlgorithm::SHA256
+        } else {
+            eyre::bail!("Unsupported digest algorithm: {:?}", value.algorithm())
+        };
+
+        Ok(Self {
+            bytes: value.as_ref().to_vec(),
+            algorithm,
+        })
+    }
+}
+
+impl std::fmt::Display for Digest {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_fmt(format_args!(
+            "{}:{}",
+            self.algorithm,
+            hex::encode(&self.bytes)
+        ))
+    }
+}
+
+impl Serialize for Digest {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        self.to_string().serialize(serializer)
+    }
+}
+
+struct DigestVisitor;
+
+impl<'de> Visitor<'de> for DigestVisitor {
+    type Value = Digest;
+
+    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+        formatter.write_str("a hexadecimal encoded cryptographic digest")
+    }
+
+    fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        let mut parts = value.split(':');
+        let algorithm_tag = parts.next().ok_or(E::missing_field("algorithm"))?;
+        let algorithm = algorithm_tag
+            .parse::<DigestAlgorithm>()
+            .map_err(|_| E::custom("invalid digest algorithm"))?;
+        let bytes = parts
+            .next()
+            .ok_or(E::missing_field("bytes"))
+            .and_then(|h| hex::decode(h).map_err(|_| E::custom("invalid encoding")))?;
+        Ok(Self::Value { algorithm, bytes })
+    }
+}
+
+impl<'de> Deserialize<'de> for Digest {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        deserializer.deserialize_str(DigestVisitor)
+    }
+}
+
+impl PartialEq for Digest {
+    fn eq(&self, other: &Self) -> bool {
+        self.algorithm == other.algorithm && self.bytes == other.bytes
+    }
+}
+
+impl Eq for Digest {}
+
+impl Ord for Digest {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.bytes.cmp(&other.bytes)
```

Comment:
Why doesn't this have to take into account the algorithm?

Reply:
In practice all we care about is the data since we only use one algorithm (and even if we introduce another one in the future, I don't think we'd ever mix algorithm types together). If we ever introduce other supported algorithms we can update this to have some sort of artificial ordering between algorithm types, just to make it extra robust, but for now it's literally pointless as we'd be comparing Self::SHA256 with itself.

Reply:
My concern is more-so that if you _do_ introduce other digest algorithms, you almost certainly won't then remember to update this `impl`. So why not just have it take into account digest from the beginning (same way `Eq` does)?

Reply:
Ok, after some sleep I figured I can implement a trivial cmp for Algorithm and that will leave this implementation here stable against future additions. Plus, the other cmp implementation will be sensitive to new additions so no one will forget to update it.